code/modules/clothing/masks/gas_filter.dm

```diff
@@ -0,0 +1,113 @@
+#define LOW_FILTERING_MOLES 0.0005
+///Min amount of wear that the filter gets when used
+#define FILTERS_CONSTANT_WEAR 0.05
+
+/obj/item/gas_filter
+	name = "atmospheric gas filter"
+	desc = "A piece of filtering cloth to be used with atmospheric gas masks and emergency gas masks."
+	icon = 'icons/obj/clothing/masks.dmi'
+	icon_state = "gas_atmos_filter"
+	///Amount of filtering points available
+	var/filter_status = 100
+	///Strenght of the filter against high filtering gases
+	var/filter_strenght_high = 10
+	///Strenght of the filter against mid filtering gases
+	var/filter_strenght_mid = 8
+	///Strenght of the filter against low filtering gases
+	var/filter_strenght_low = 5
+	///General efficiency of the filter (between 0 and 1)
+	var/filter_efficiency = 0.5
+	///All the gases that are breathed and to be checked against
+	var/list/gases_moles
+
+/obj/item/gas_filter/Initialize()
+	. = ..()
+	for(var/gas_id in GLOB.meta_gas_info)
+		LAZYADD(gases_moles, gas_id)
+
+/obj/item/gas_filter/examine(mob/user)
+	. = ..()
+	. += "<span class='notice'>[src] is at <b>[filter_status]%</b> durability.</span>"
+
+/**
+ * called by the gas mask where the filter is installed, lower the filter_status depending on the breath gas composition and by the strenght of the filter
+ * returns the modified breath gas mixture
+ *
+ * Arguments:
+ * * breath - the current gas mixture of the breathed air
+ *
+ */
+/obj/item/gas_filter/proc/reduce_filter_status(datum/gas_mixture/breath)
+
+	for(var/gas_id in GLOB.meta_gas_info)
+		gases_moles[gas_id] = breath.gases[gas_id][MOLES]
+
+	var/danger_points = 0
+
+	for(var/gas_id in GLOB.high_filtering_gases)
+		if(gases_moles[gas_id] > HIGH_FILTERING_MOLES)
+			breath.gases[gas_id][MOLES] = max(breath.gases[gas_id][MOLES] - filter_strenght_high * filter_efficiency * HIGH_FILTERING_RATIO, 0)
+			danger_points += 0.5
+		else if(gases_moles[gas_id] > 0)
+			breath.gases[gas_id][MOLES] = max(breath.gases[gas_id][MOLES] - filter_strenght_high * filter_efficiency * LOW_FILTERING_RATIO, 0)
+			danger_points += 0.05
+
+	for(var/gas_id in GLOB.mid_filtering_gases)
+		if(gases_moles[gas_id] > MID_FILTERING_MOLES)
+			breath.gases[gas_id][MOLES] = max(breath.gases[gas_id][MOLES] - filter_strenght_mid * filter_efficiency * HIGH_FILTERING_RATIO, 0)
+			danger_points += 0.75
+		else if(gases_moles[gas_id] > 0)
+			breath.gases[gas_id][MOLES] = max(breath.gases[gas_id][MOLES] - filter_strenght_mid * filter_efficiency * LOW_FILTERING_RATIO, 0)
+			danger_points += 0.15
+
+	for(var/gas_id in GLOB.low_filtering_gases)
+		if(gases_moles[gas_id] > LOW_FILTERING_MOLES)
+			breath.gases[gas_id][MOLES] = max(breath.gases[gas_id][MOLES] - filter_strenght_low * filter_efficiency * HIGH_FILTERING_RATIO, 0)
+			danger_points += 1
+		else if(gases_moles[gas_id] > 0)
+			breath.gases[gas_id][MOLES] = max(breath.gases[gas_id][MOLES] - filter_strenght_low * filter_efficiency * LOW_FILTERING_RATIO, 0)
+			danger_points += 0.5
```

Comment:
Wouldn't it be better to do one for loop over the contents of breath.gases, and then sort it into the  different filtering categories from there?

Reply:
will do